blank_react_project/src/App.js

```diff
@@ -60,7 +62,8 @@ class App extends Component {
   }
 
   async changeGreeting() {
-    await this.props.contract.setGreeting({ message: 'howdy' });
+    this.greetingIndex = (this.greetingIndex + 1) % 3;
```

Comment:
Same thing here – `greetings.length`

And here, we can similarly move `greetings` and `greetingIndex` outside of the component.

```js
const greetings = ['hello, 'aloha', 'bonjour'];
var greetings = 0

class App extends Component {
...
```

Reply:
But wait! Here and above, we're not initializing `greeting` from the value in the contract! If you refresh the page, you will see "Hello" again, even if the contract state says "Aloha" – right?

Reply:
> But wait! Here and above, we're not initializing `greeting` from the value in the contract! If you refresh the page, you will see "Hello" again, even if the contract state says "Aloha" – right?

Not sure what you mean, there is no "Hello" in the frontend at all. What happened when you tried to refresh it?

Reply:
You're not using the value returned by `welcome()` to figure out the initial `greetingIndex`. That means that your greeting index might not match what's stored on-chain. This leads to two problems:

1. if you refresh the page when the contract is already storing "aloha", then clicking the button will not do anything
2. now that this frontend logic is more complex, it adds confusing coupling between the frontend & backend; I think it makes this app code harder to understand